src/tof/result.py

```diff
@@ -159,13 +161,106 @@ def __getitem__(self, name: str) -> Union[ChopperReading, DetectorReading]:
 
+    def _plot_visible_rays(
+        self,
+        max_rays: int,
+        pulse_index: int,
+        furthest_detector: DetectorReading,
+        ax: plt.Axes,
+        cbar: bool,
+        wmin: sc.Variable,
+        wmax: sc.Variable,
+    ):
+        if max_rays <= 0:
+            return
+        da = furthest_detector.data['pulse', pulse_index]
+        visible = da[~da.masks['blocked_by_others']]
+        tofs = visible.coords['tof']
+        if (max_rays is not None) and (len(tofs) > max_rays):
+            inds = np.random.choice(len(tofs), size=max_rays, replace=False)
+        else:
+            inds = slice(None)
+        birth_times = visible.coords['time'][inds]
+        wavelengths = visible.coords['wavelength'][inds]
+        distances = furthest_detector.distance.broadcast(sizes=birth_times.sizes)
+        _add_rays(
+            ax=ax,
+            tofs=tofs[inds],
+            birth_times=birth_times,
+            distances=distances,
+            cbar=cbar,
+            wavelengths=wavelengths,
+            wmin=wmin,
+            wmax=wmax,
+        )
+
+    def _plot_blocked_rays(
+        self,
+        blocked_rays: int,
+        pulse_index: int,
+        furthest_detector: DetectorReading,
+        ax: plt.Axes,
+    ):
+        if blocked_rays <= 0:
+            return
+        slc = ('pulse', pulse_index)
+        inv_mask = ~self._masks[furthest_detector.name][slc]
+        nrays = int(inv_mask.sum())
+        if nrays > blocked_rays:
+            inds = np.random.choice(nrays, size=blocked_rays, replace=False)
+        else:
+            inds = slice(None)
+        birth_times = self._source.data.coords['time'][slc][inv_mask][inds]
+
+        components = sorted(
+            chain(self._choppers.values(), [furthest_detector]),
+            key=lambda c: c.distance.value,
+        )
+        dim = 'component'
+        tofs = sc.concat(
+            [
+                self._arrival_times[comp.name][slc][inv_mask][inds]
+                for comp in components
+            ],
+            dim=dim,
+        )
+        distances = sc.concat(
+            [comp.distance.broadcast(sizes=birth_times.sizes) for comp in components],
+            dim=dim,
+        )
+        masks = sc.concat(
+            [sc.ones(sizes=birth_times.sizes, dtype=bool)]
+            + [self._masks[comp.name][slc][inv_mask][inds] for comp in components],
+            dim=dim,
+        )
+
+        diff = sc.abs(masks[dim, 1:].to(dtype=int) - masks[dim, :-1].to(dtype=int))
+        diff.unit = ''
+        _add_rays(
+            ax=ax,
+            tofs=(tofs * diff).max(dim=dim),
+            birth_times=birth_times,
+            distances=(distances * diff).max(dim=dim),
+        )
+
+    def _plot_pulse(self, pulse_index: int, ax: plt.Axes):
+        time_coord = self.source.data.coords['time']['pulse', pulse_index]
+        tmin = time_coord.min().to(unit='us').value
```

Comment:
It is a bit odd to have the unit conversions every. Have you considered converting once,, when setting the time coords in the source or components?